runtime/src/unchecked_extrinsic.rs

```diff
@@ -0,0 +1,437 @@
+	for UncheckedExtrinsic<Address, Call, Extra>
+where
+	Address: Member + MaybeDisplay,
+	Call: Encode + Member,
+	Extra: SignedExtension<AccountId = AccountId>,
+	AccountId: Member + MaybeDisplay + Into<AccountId32>,
+	Lookup: traits::Lookup<Source = Address, Target = AccountId>,
+{
+	type Checked = CheckedExtrinsic<AccountId, Call, Extra>;
+
+	fn check(self, lookup: &Lookup) -> Result<Self::Checked, TransactionValidityError> {
+		Ok(match self.signature {
+			Some((signed, signature, extra)) => {
+				let signed = lookup.lookup(signed)?;
+				let raw_payload = SignedPayload::new(self.function, extra)?;
+				let custom_signature = CustomSignature { signature };
+				if !raw_payload.using_encoded(|payload| {
+					custom_signature.verify(payload, &signed.clone().into())
+				}) {
+					return Err(InvalidTransaction::BadProof.into())
+				}
+
+				let (function, extra, _) = raw_payload.deconstruct();
+				CheckedExtrinsic { signed: Some((signed, extra)), function }
+			},
+			None => CheckedExtrinsic { signed: None, function: self.function },
+		})
+	}
+}
+
+impl<Address, Call, Extra> ExtrinsicMetadata for UncheckedExtrinsic<Address, Call, Extra>
+where
+	Extra: SignedExtension,
+{
+	const VERSION: u8 = EXTRINSIC_FORMAT_VERSION;
+	type SignedExtensions = Extra;
+}
+
+impl<Address, Call, Extra> Decode for UncheckedExtrinsic<Address, Call, Extra>
+where
+	Address: Decode,
+	Call: Decode,
+	Extra: SignedExtension,
+{
+	fn decode<I: Input>(input: &mut I) -> Result<Self, Error> {
+		// This is a little more complicated than usual since the binary format must be compatible
+		// with SCALE's generic `Vec<u8>` type. Basically this just means accepting that there
+		// will be a prefix of vector length.
+		let expected_length: Compact<u32> = Decode::decode(input)?;
+		let before_length = input.remaining_len()?;
+
+		let version = input.read_byte()?;
+
+		let is_signed = version & 0b1000_0000 != 0;
+		let version = version & 0b0111_1111;
+		if version != EXTRINSIC_FORMAT_VERSION {
+			return Err("Invalid transaction version".into())
+		}
+
+		let signature = is_signed.then(|| Decode::decode(input)).transpose()?;
+		let function = Decode::decode(input)?;
+
+		if let Some((before_length, after_length)) =
+			input.remaining_len()?.and_then(|a| before_length.map(|b| (b, a)))
+		{
+			let length = before_length.saturating_sub(after_length);
+
+			if length != expected_length.0 as usize {
+				return Err("Invalid length prefix".into())
+			}
+		}
+
+		Ok(Self { signature, function })
+	}
+}
+
+impl<Address, Call, Extra> Encode for UncheckedExtrinsic<Address, Call, Extra>
+where
+	Address: Encode,
+	Call: Encode,
+	Extra: SignedExtension,
+{
+	fn encode(&self) -> Vec<u8> {
+		let mut tmp = Vec::with_capacity(sp_std::mem::size_of::<Self>());
+
+		// 1 byte version id.
+		match self.signature.as_ref() {
+			Some(s) => {
+				tmp.push(EXTRINSIC_FORMAT_VERSION | 0b1000_0000);
+				s.encode_to(&mut tmp);
```

Comment:
Are we ok that allocated tmp will be actually smaller if 's' in None?

Reply:
I think it will fine.